src/server/pfs/drive/driver.go

```diff
@@ -217,25 +220,44 @@ func (d *driver) StartCommit(repo *pfs.Repo, commitID string, parentID string, b
 			return err
 		}
 	}
+	d.commitConds[commitID] = sync.NewCond(&d.lock)
 	return nil
 }
 
-func (d *driver) FinishCommit(commit *pfs.Commit, finished *google_protobuf.Timestamp, shards map[uint64]bool) error {
+// FinishCommit blocks until its parent has been finished/cancelled
+func (d *driver) FinishCommit(commit *pfs.Commit, finished *google_protobuf.Timestamp, cancelled bool, shards map[uint64]bool) error {
+	canonicalCommit, err := d.canonicalCommit(commit)
+	if err != nil {
+		return err
+	}
 	// closure so we can defer Unlock
 	var diffInfos []*pfs.DiffInfo
 	if err := func() error {
 		d.lock.Lock()
 		defer d.lock.Unlock()
-		canonicalCommit, err := d.canonicalCommit(commit)
-		if err != nil {
-			return err
-		}
 		for shard := range shards {
 			diffInfo, ok := d.diffs.get(pfsclient.NewDiff(canonicalCommit.Repo.Name, canonicalCommit.ID, shard))
 			if !ok {
 				return fmt.Errorf("commit %s/%s not found", canonicalCommit.Repo.Name, canonicalCommit.ID)
 			}
+			if diffInfo.ParentCommit != nil {
+				parentDiffInfo, ok := d.diffs.get(pfsclient.NewDiff(canonicalCommit.Repo.Name, diffInfo.ParentCommit.ID, shard))
+				if !ok {
+					return fmt.Errorf("parent commit %s/%s not found", canonicalCommit.Repo.Name, diffInfo.ParentCommit.ID)
+				}
+				// Wait for parent to finish
+				for parentDiffInfo.Finished == nil {
+					cond, ok := d.commitConds[diffInfo.ParentCommit.ID]
+					if !ok {
+						return fmt.Errorf("parent commit %s/%s was not finished but a corresponding conditional variable could not be found; this is likely a bug")
+					}
+					cond.Wait()
```

Comment:
I'm not quite sure if this works. According to the documentation for `CondVar` there's no guarantee that the condition is true when `Wait` returns and you need to [wrap it in a loop](https://golang.org/pkg/sync/#Cond.Wait).

I generally like to avoid CondVars, mostly because the `Wait` semantics are so weird. I think what you're doing here would probably be better done with a simple channel. CondVars are useful when you have a condition that can switch between true and false repeatedly, but commits only get finished once so you can basically swap `close(c)` for `Broadcast` and `<-c` for `Wait` and get exactly the semantics you want.